internal/core/future.rs

```diff
@@ -0,0 +1,163 @@
+use core::task::Poll;
+
+enum FutureState<T> {
+    Running(Pin<Box<dyn Future<Output = T>>>),
+    Finished(Option<T>),
+}
+
+struct FutureRunnerInner<T> {
+    fut: FutureState<T>,
+    wakers: Vec<core::task::Waker>,
+}
+
+struct FutureRunner<T> {
+    #[cfg(not(feature = "std"))]
+    inner: core::cell::RefCell<FutureRunnerInner<T>>,
+    #[cfg(feature = "std")]
+    inner: std::sync::Mutex<FutureRunnerInner<T>>,
+    aborted: atomic::AtomicBool,
+
+    #[cfg(feature = "std")]
+    thread: std::thread::ThreadId,
+}
+
+impl<T> FutureRunner<T> {
+    fn inner(&self) -> impl DerefMut<Target = FutureRunnerInner<T>> + '_ {
+        #[cfg(feature = "std")]
+        return self.inner.lock().unwrap();
+        #[cfg(not(feature = "std"))]
+        return self.inner.borrow_mut();
+    }
+}
+
+// # Safety:
+// The Future might not be Send, but we only poll the future from the main thread.
+// (We even assert that)
+// We may access the finished value from another thread only if T is Send
+// (because JoinHandle only implement Send if T:Send)
+#[allow(unsafe_code)]
+unsafe impl<T> Send for FutureRunner<T> {}
+#[allow(unsafe_code)]
+unsafe impl<T> Sync for FutureRunner<T> {}
+
+impl<T: 'static> Wake for FutureRunner<T> {
+    fn wake(self: alloc::sync::Arc<Self>) {
+        invoke_from_event_loop(move || {
+            #[cfg(feature = "std")]
+            assert_eq!(self.thread, std::thread::current().id(), "the future was moved to a thread despite we checked it was created in the event loop thread");
+            let waker = self.clone().into();
+            let mut inner = self.inner();
+            let mut cx = core::task::Context::from_waker(&waker);
+            if let FutureState::Running(fut) = &mut inner.fut {
+                if self.aborted.load(atomic::Ordering::Relaxed) {
+                    inner.fut = FutureState::Finished(None);
+                } else {
+                    match fut.as_mut().poll(&mut cx) {
+                        Poll::Ready(val) => inner.fut = FutureState::Finished(Some(val)),
+                        Poll::Pending => {}
+                    }
+                }
+            }
+        })
+        .expect("No event loop despite we checked");
+    }
+}
+
+/// The return value of the [`spawn_future()`] function
+///
+/// Can be used to abort the future, or to get the value from a different thread with `.await`
+///
+/// This trait implements future. Polling it after it finished or aborted may result in a panic.
+pub struct JoinHandle<T>(alloc::sync::Arc<FutureRunner<T>>);
+
+impl<T> Future for JoinHandle<T> {
+    type Output = T;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut core::task::Context<'_>) -> Poll<Self::Output> {
+        let mut inner = self.0.inner();
+        match &mut inner.fut {
+            FutureState::Running(_) => {
+                let waker = cx.waker();
+                if !inner.wakers.iter().any(|w| w.will_wake(waker)) {
+                    inner.wakers.push(waker.clone());
```

Comment:
It seems like to me that these wakers are never woken up.

Reply:
well spotted!  They were meant to be woken in FutureRunner::wake.  I wonder why the test pass.